### PR TITLE
[SYCL] Fix memory leak for sycl::device::get_devices

### PR DESCRIPTION
**Pull request by bso-intel**
Plugins create pi_device objects for root devices, but
it does not call piDeviceRelease and causes memory leaks.
This slows the performance of SYCL apps.

Signed-off-by: Byoungro So <byoungro.so@intel.com>

### CODE REVIEW

**Comment by bso-intel** [comment]
@bader @vladimirlaz @romanovvlad 
Sorry, I had to create a new branch to show only my own changes.
I accommodated all your feedback in the previous PR. https://github.com/intel/llvm/pull/1923

**Comment by bso-intel** [comment]
@bader, @DoyleLi ,
It seems the failures reported in this PR are already failing in the sycl branch.
I tested with the latest head of sycl branch, and it also failed on the group-algorithm/barrier.cpp.
So, I think these failures are not caused by my change.
Could you confirm if this is true?

File: sycl/test/basic_tests/memory-consumption.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+//
+// UNSUPPORTED: windows
+//
+//==-----memory-consumption.cpp - SYCL memory consumption basic test ------==//
```
Comment (romanovvlad) [inline sycl/test/basic_tests/memory-consumption.cpp:6]: I think this test is valuable to check overall memory consumption. But I suggest adding more specific test on the change you are doing, such a verification could be done by setting SYCL_PI_TRACE and checking its output(see sycl/test/scheduler/ReleaseResourcesTest.cpp as an example).

Comment (bso-intel) [reply sycl/test/basic_tests/memory-consumption.cpp:6]: I will make a follow-up patch for this SYCL_PI_TRACE later.

Comment (romanovvlad) [reply sycl/test/basic_tests/memory-consumption.cpp:6]: I'm OK if additional test will be added in the follow up patch.

**Pull request merged**